Automattic/Gravatar-SDK-Android
# QuickEditor: Avatar image download option

Closes #447

### Description

There's more than one way to solve this task. But there's no solution (that I know of) that would drop the requirement for the `WRITE_EXTERNAL_STORAGE` permission on API 28 and lower (pre-SAF).

1. We could handle the download ourselves (to a temp private directory) and then request to create a file in a public one [via Intent](https://developer.android.com/training/data-storage/shared/documents-files#create-file). 

This would require from us to do all the heavy work of showing notifications, handling retries, etc. We would likely need extra dependency for this (WorkManager), so that wouldn't be ideal. On top of that, the flow would be longer, as the user would have to pick the directory for the file to be saved. Although some would prefer that, in general, I think it's enough to save it in `Downloads` folder. 

2. Use `DownloadManager`.

This still requires the `WRITE_EXTERNAL_STORAGE` on API 28 and lower, but it handles the download, notifications, and retries. The flow on AP I29+ is extremely simple as there are no extra actions required from the user. No extra dependency as well.

There are some cons. 
- The user might disable the `DownloadManager`. Thankfully, we can detect that and direct the user to a proper settings screen to enable it.
- We don't have full control, but we can listen for updates via `BroadcastReceiver`. For now, I've skipped listening for updates. There's a notification added once the download is completed, and I didn't see much value in duplicating that in the QE via Snackbar. 

I've decided to pick option nr 2 as it's simple and offers a quicker flow for the user. We can always evaluate when we encounter issued with the `DownloadManager`.
 
| API 28 | API 28 Permission rationale dialog | API 29+ | 
|---|---|---|
| ![download_api28](https://github.com/user-attachments/assets/3a37ada6-18f5-4674-b86b-bf68ab828f2c) | <img width="404" alt="Screenshot 2024-11-21 at 11 29 08" src="https://github.com/user-attachments/assets/dcc3fb84-84a7-4a1e-82bb-5ca3d9141a60"> | ![download_api_29+](https://github.com/user-attachments/assets/a684d1c5-0bdf-4ae1-91cb-4aad4ec35613) |

### Testing Steps

1. Launch the QE on API 29+
2. Tap on the`...` button on one Avatar
3. Tap Download image
4. Confirm that a Snackbar with information about queued download was shown
5. Check the notification for the image download

Repeat on a device with API 28 or lower. Verify the flow with granted and denied permission.

`DownloadManager` disabled scenario:

1. Go to System settings -> Apps  -> All Apps -> Menu -> Show system apps
2. Look for DownloadManager
3. Disable it
4. Go back to QE and try downloading the avatar
